// gulp-lego/lib/transport.js
'use strict';

var extname = require('path').extname;
var join = require('path').join;
var extend = require('extend');
var mixarg = require('mixarg');
var through = require('through2');
var duplexer2 = require('duplexer2');
var pipe = require('multipipe');
var gulpswitch = require('stream-switch');
var gulpif = require('gulp-if');
var reactify = require('gulp-reactify');
var is = require('is-type');
var Package = require('father').LegoPackage;
var standalonify = require('lego-standalonify');
var plugin = require('./plugin');
var include = plugin.include;
var concat = plugin.concat;
var js = plugin.js;
var css = plugin.css;
var css2js = plugin.css2js;
var json = plugin.json;
var tpl = plugin.tpl;
var html = plugin.html;
var handlebars = plugin.handlebars;

var defaults = {
  // for parse
  cwd: process.cwd(),
  moduleDir: 'lego_modules',
  skip: '',
  ignore: '',

  // for transport
  include: 'relative',
  global: '',
  idleading: '{{name}}/{{version}}',
  rename: '',
  stream: null,
  pkg: null,
  standalone: null
};

module.exports = function(opt) {
  var cwd = opt.cwd || defaults.cwd;
  var pkg = require(join(cwd, 'package.json'));
  var lego = pkg.lego || {};

  opt = mixarg(defaults, lego.buildArgs || '', opt);
  if (is.string(opt.ignore)) opt.ignore = opt.ignore ? opt.ignore.split(/\s*,\s*/) : [];
  if (is.string(opt.skip)) opt.skip = opt.skip ? opt.skip.split(/\s*,\s*/) : [];
  opt.global = getGlobal(opt.global);
  opt.skip = opt.skip.concat(Object.keys(opt.global));
  if (!opt.pkg) {
    opt.pkg = new Package(opt.cwd, {
      skip: opt.skip,
      ignore: opt.ignore,
      moduleDir: opt.moduleDir
    });
  }

  var inputStream = through.obj();
  var outputStream = through.obj();

  var streams = getStream(opt);
  var isStandalone = opt.include === 'standalone';
  var isUmd = opt.include === 'umd';
  if (isUmd) {
    opt.standalone = extend({umd: camelCase(opt.pkg.name)}, opt.standalone);
  }

  var jsStream = pipe(
    include(opt),
    parseByType(streams),
    reactify(),
    streams.js,
    concat(opt),
    gulpif(isStandalone || isUmd, standalonify(opt.standalone))
  );
  var cssStream = streams.css;

  inputStream
  .pipe(plugin.file(opt.pkg))
  .pipe(chooseParseType(jsStream, cssStream))
  .pipe(plugin.dest(opt))
  .pipe(outputStream);

  errorHandle(jsStream);
  errorHandle(cssStream);

  return duplexer2(inputStream, outputStream);

  function errorHandle(stream) {
    stream.on('error', function(e) {
      outputStream.emit('error', e);
    });
  }
};

function getStream(opt) {
  var defaultStream = {
    '.css': css(opt),
    '.css.js': pipe(
      // overide rename for this
      css(extend({}, opt, {rename: rename})),
      css2js(opt)
    ),
    '.json': json(opt),
    '.tpl': tpl(opt),
    '.html': html(opt),
    '.js': js(opt),
    '.handlebars': handlebars(opt)
  };

  var stream = opt.stream || {}, ret = {other: {}};
  Object.keys(defaultStream).forEach(function(key) {
    var func = stream[key];
    if (func && typeof func !== 'function') {
      throw new Error('opt.stream\'s value should be function');
    }
    var val = (typeof func === 'function' && func(opt)) || defaultStream[key];
    if (key === '.css' || key === '.js') {
      ret[key.substring(1)] = val;
    } else {
      ret.other[key] = val;
    }
  });

  return ret;

  function rename(file) {
    return file;
  }
}

function chooseParseType(jsStream, cssStream) {
  return gulpswitch(function(file) {
    var ext = extname(file.path);
    switch(ext) {
      case '.js':
      case '.json':
      case '.tpl':
      case '.handlebars':
        return '.js';
      case '.css':
        return '.css';
    }
  }, {
    '.js': jsStream,
    '.css': cssStream
  });
}

function parseByType(streams) {
  return gulpswitch(function switchCondition(file) {
    var ext = extname(file.path);
    return ext === '.css' ? '.css.js' : ext;
  }, streams.other);
}

function camelCase(str) {
  return str.replace(/[_.-](\w|$)/g, function (_,x) {
    return x.toUpperCase();
  });
}

function getGlobal(str) {
  if (({}).toString.call(str) === '[object Object]') return str;
  if (typeof str !== 'string') return {};

  var ret = {};
  str.split(/\s*,\s*/).forEach(function(item) {
    var m = item.split(':');
    ret[m[0]] = m[1];
  });
  return ret;
}
